virtual_rainforest/models/hydrology/hydrology_model.py

```diff
@@ -224,24 +239,68 @@ def setup(self) -> None:
             )
         )
 
+        # Turn elevation data into np array
+        elevation = np.array(self.data["elevation"])
+
+        # map neighbours and identify downstream cell_ids
+        mapping = [
+            self.data.grid.neighbours[cell_id]
+            for cell_id in range(len(self.data.grid.cell_id))
+        ]
+        downstream_ids = []
+
+        for cell_id, neighbors_id in enumerate(mapping):
+            # Find lowest neighbour
+            downstream_id_loc = np.argmax(elevation[cell_id] - elevation[neighbors_id])
+            downstream_ids.append(neighbors_id[downstream_id_loc])
+
+        # Invert mapping of cell_id: downstream neighbour to cell_id: upstream_neighbors
+        upstream_ids: list = [[] for i in range(len(self.data.grid.cell_id))]
+        for down_s, up_s in enumerate(downstream_ids):
+            upstream_ids[up_s].append(down_s)
+
+        self.upstream_ids = upstream_ids
+        """Upstream ID's for the calculation of accumulated surface runoff."""
```

Comment:
I'm wondering if the DEM / runoff / stream flow should all be part of the setup of `core.grid.Grid` so it can be initialised right at the start and then shared? 

That doesn't need doing right now - and it is all quite specific to hydrology.  But @TaranRallings I can imagine we might want a graph showing the relative heights of cells for calculating movement costs at some point? That would use a lot of the same information.

I definitely agree that this should become a separate function - that would also make it easy to move into `core.grid.Grid` in the future (possibly as one of a range of runoff models).

Reply:
I agree. I separated the functions and added tests. I also changed the variable names slightly to make in clearer what they represent.

Reply:
There was initially a comment that this should all move to core which might have gone lost somewhere, but I totally agree. I put it here for easy testing but I'm very happy to move this to core when it is the right time :-)